### PR TITLE
The mother of all pull requests: add --incremental

### PR DESCRIPTION
**Pull request by gvanrossum**
tl;dr: Please review in this order: semanal.py+tests; nodes.py+types.py; fixup.py; build.py; main.py.

See https://paper.dropbox.com/doc/Mypy-incremental-checking-ifj77IF7eNNySxhXYAaBe#:h=Technical-debt for some technical debt (read the paper from the top for a more complete overview).

Despite that technical debt I'd like to merge this into master in the current state, because at this point it works well enough and in particular I've tested it with typeshed's runtests.py script as well as with the selection of files in the internal Dropbox server repo. Some considerations:

- The new (SCC + topsort) algorithm for determining dependencies is always used
- The incremental (cache) feature is only used when `-i/--incremental` is used
- After merging this would require additional real-world testing before we can do a release
- My own biggest concern is the hack for the sqlalchemy cycle: https://github.com/gvanrossum/mypy/commit/ede81849a636e88d4f048472bc290224aa0cc7a0

Don't be fooled by the large number of commits! There are a lot of dead ends and U-turns that you can ignore. Most of the changes are limited to a small number of files:

- build.py: the new dependency manager (I ripped out the old one)
- nodes.py, types.py: added [de]serialize() methods to many classes
- fixup.py: new module to fix up symbol tables after deserialization
- main.py: added the `-i/--incremental` option
- semanal.py: a fix for anonymous namedtuples needed by serialization
- some tests added/changed for the namedtuple fix
- .gitignore the .mypy_cache directory (though not even sure I want this)

Somehow GitHub refuses to show the diffs for build.py. Sorry about this; I guess you have to view them locally. You could also try to view the incremental diffs to this file starting with https://github.com/python/mypy/commit/0b4ef76db49c0cb3719ef7d3c6688bcfc2c81b3a. (Also note that the dates GitHub lists for the diffs are bogus -- they represent the most recent rebase. The actual work started over a month ago.)

Some questions about how to merge this monster into master. I've got a feeling that the detailed commits from my branch are not that useful. I propose the following order, as separate commits to master:

1. the semanal.py changes and related tests
2. (de)serialization methods in nodes.py, types.py
3. the rest: fixup.py, build.py, main.py, .gitignore

This may also represent a reasonable review order.

### CODE REVIEW

**Comment by JukkaL** [comment]
Thanks for implementing all this! I'll look at this later this week -- I may need to allocate some time on my calendar to go through this all :-)

**Comment by JukkaL** [comment]
Some random ideas about breaking cycles below.

We could perhaps rank dependencies based on some general criteria, and give preference to the higher ranks when sorting within a SCC. For example, `from m import x` at module top level would get the highest rank, as the target module will have to be initialized for the statement to succeed, so it always indicates a "strong" dependency. The next rank would include `import x` at module top level -- these can be evaluated "both ways" when there are cyclic imports. The lowest rank would be given to all imports within functions, as some of these might only exist to work around cycles. The intuition is that this might result in a somewhat "natural" dependency ordering when there are cycles.

We could also use your observation to give a higher rank to intra-package dependencies.

**Comment by JukkaL** [comment]
Also, if module `x` only imports module `y` because of a type annotation in `x` uses a type defined in `y`, we could give this dependency a low rank, as this dependency wouldn't exist without type annotations. However, not sure if this would actually be helpful, as the dependency is quite real from a type checker's perspective.

**Comment by ddfisher** [comment]
My inclination for how to approach the cycle issue would be to start by investigating which problems arise when files are processed in the wrong order.  Once we have some understanding of why these different problems occur, we can start thinking about ways to heuristically fix them (or to introduce additional type checking steps).

**Comment by JukkaL** [comment]
@ddfisher  That is a good suggestion, though we already know some things that will cause trouble, such as class and module level attributes potentially not having inferred types, and decorated functions not getting inferred types. Another potential problem is that the ordering within a cycle may change across runs, resulting in inconsistent type checking results / errors. It would be nice to make the ordering at least mostly stable with respect to most normal edits, including things like renaming a module. We might also want to do multi pass type checking across a dependency cycle, but that's a different can of worms.

Another idea for ranking: if a class in module `x` subclasses a class defined in module `y`, this is a super strong dependency and we always want to type check `y` before `x`.

**Comment by gvanrossum** [comment]
I'd like to hold off on any advanced algorithms. But there are two things
I'd like to try.

- First stabilize the order. There's one source of randomness (due to
Python's hash randomization): topsort() yields ready items in a randomized
order. This could easily be stabilized by yielding the set of ready items
and in the caller (sorted_components()) sort them by reverse State.order
(probably the min or max order within each scc).

- Second, try David's idea of modifying the handling of "from a import b"
in all_imported_modules_in_file() to only add a dependency on a if b is
*not* a module. This alone made the sqlalchemy problem resolve beautifully
without any hacks in process_graph(). It caused a randomly appearing and
disappearing error in the server code but I think the previous bullet will
avoid that.

On Wed, Mar 16, 2016 at 12:12 PM, Jukka Lehtosalo <notifications@github.com>
wrote:

> @ddfisher <https://github.com/ddfisher> That is a good suggestion, though
> we already know some things that will cause trouble, such as class and
> module level attributes potentially not having inferred types, and
> decorated functions not getting inferred types. Another potential problem
> is that the ordering within a cycle may change across runs, resulting in
> inconsistent type checking results / errors. It would be nice to make the
> ordering at least mostly stable with respect to most normal edits,
> including things like renaming a module. We might also want to do multi
> pass type checking across a dependency cycle, but that's a different can of
> worms.
>
> Another idea for ranking: if a class in module x subclasses a class
> defined in module y, this is a super strong dependency and we always want
> to type check y before x.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/python/mypy/pull/1292#issuecomment-197495928>
>



-- 
--Guido van Rossum (python.org/~guido)

**Comment by JukkaL** [comment]
Agreed, I'm not suggesting to implement anything fancy right now, but there are some ideas I had for potential future things.

David's suggestion is great -- we should definitely do that.

**Comment by gvanrossum** [comment]
OK, pushed both things (stabilized order and David's idea). Please review.

File: mypy/fixup.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+                        LDEF, MDEF, GDEF, MODULE_REF)
+from mypy.types import (CallableType, EllipsisType, Instance, Overloaded, TupleType,
+                        TypeList, TypeVarType, UnboundType, UnionType, TypeVisitor)
+from mypy.visitor import NodeVisitor
+
+
+def fixup_module_pass_one(tree: MypyFile, modules: Dict[str, MypyFile]) -> None:
+    node_fixer = NodeFixer(modules)
+    node_fixer.visit_symbol_table(tree.names)
+    # print('Done pass 1', tree.fullname())
```
Comment (JukkaL) [inline mypy/fixup.py]: Commented out print statements. If these are useful, it's probably better to hind them behind a debug flag.

File: mypy/fixup.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+
+def fixup_module_pass_two(tree: MypyFile, modules: Dict[str, MypyFile]) -> None:
+    compute_all_mros(tree.names, modules)
+    # print('Done pass 2', tree.fullname())
```
Comment (JukkaL) [inline mypy/fixup.py]: Another print statement.

File: mypy/fixup.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+        if value.kind in (LDEF, MDEF, GDEF) and isinstance(value.node, TypeInfo):
+            info = value.node
+            # print('  Calc MRO for', info.fullname())
+            try:
+                info.calculate_mro()
+            except Exception:
+                import pdb
+                pdb.set_trace()
+            if not info.mro:
+                print('*** No MRO calculated for', info.fullname())
```
Comment (JukkaL) [inline mypy/fixup.py]: Handle this better?

File: mypy/fixup.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+def fixup_module_pass_two(tree: MypyFile, modules: Dict[str, MypyFile]) -> None:
+    compute_all_mros(tree.names, modules)
+    # print('Done pass 2', tree.fullname())
+
+
+def compute_all_mros(symtab: SymbolTable, modules: Dict[str, MypyFile]) -> None:
+    for key, value in symtab.items():
+        if value.kind in (LDEF, MDEF, GDEF) and isinstance(value.node, TypeInfo):
+            info = value.node
+            # print('  Calc MRO for', info.fullname())
```
Comment (JukkaL) [inline mypy/fixup.py]: Another print statement.

File: mypy/fixup.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+            try:
+                info.calculate_mro()
+            except Exception:
+                import pdb
```
Comment (JukkaL) [inline mypy/fixup.py]: Hmm maybe this shouldn't be the default behavior? If you want to keep this, add a TODO comment about removing this before turning this on by default.

File: mypy/fixup.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+        if type_fixer is None:
+            type_fixer = TypeFixer(self.modules)
+        self.type_fixer = type_fixer
+
+    # NOTE: This method isn't (yet) part of the NodeVisitor API.
+    def visit_type_info(self, info: TypeInfo) -> None:
+        save_info = self.current_info
+        try:
+            self.current_info = info
+            # print('Descending into', info.fullname())
```
Comment (JukkaL) [inline mypy/fixup.py]: More print statements.

File: mypy/fixup.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+        stnode = names[key]
+        if not rest:
+            return stnode
+        node = stnode.node
+        assert isinstance(node, TypeInfo)
+        names = cast(TypeInfo, node).names
+
+
+def store_qualified(modules: Dict[str, MypyFile], name: str, info: SymbolNode) -> None:
+    print("store_qualified", name, repr(info))
```
Comment (JukkaL) [inline mypy/fixup.py]: Print statement.

File: mypy/fixup.py [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+            rest.append(tail)
+            break
+    names = mod.names
+    while True:
+        if not rest:
+            print('*** Cannot find', name)
+            return None
+        key = rest.pop()
+        if key not in names:
+            print('*** Cannot find', key, 'for', name)
```
Comment (JukkaL) [inline mypy/fixup.py]: Print statement.

File: mypy/fixup.py [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+    while True:
+        head, tail = head.rsplit('.', 1)
+        mod = modules.get(head)
+        if mod is not None:
+            rest.append(tail)
+            break
+    names = mod.names
+    while True:
+        if not rest:
+            print('*** Cannot find', name)
```
Comment (JukkaL) [inline mypy/fixup.py]: Print statement.

File: mypy/fixup.py [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
```
Comment (JukkaL) [inline mypy/fixup.py]: And more print statements below.

File: mypy/nodes.py [hunk 11]
```diff
@@ -109,6 +113,20 @@ def get_line(self) -> int:
     def accept(self, visitor: NodeVisitor[T]) -> T:
         raise RuntimeError('Not implemented')
 
+    # @abstractmethod  # TODO
```
Comment (JukkaL) [inline mypy/nodes.py]: Why is this commented out?

File: mypy/nodes.py [hunk 12]
```diff
@@ -174,6 +196,24 @@ def is_package_init_file(self) -> bool:
...
+    def serialize(self) -> JsonDict:
+        return {'.class': 'MypyFile',
+                '_name': self._name,
+                '_fullname': self._fullname,
+                'names': self.names.serialize(self._fullname),
+                'is_stub': self.is_stub,
+                }
+
+    @classmethod
+    def deserialize(cls, data: JsonDict) -> 'MypyFile':
```
Comment (JukkaL) [inline mypy/nodes.py]: `path` may be needed for some error messages that target other files -- I think that we have a few errors like that.

File: mypy/nodes.py [hunk 13]
```diff
@@ -383,6 +456,29 @@ def accept(self, visitor: NodeVisitor[T]) -> T:
...
         return self.info is not None and self._name == '__init__'
 
+    def serialize(self) -> JsonDict:
+        return {'.class': 'FuncDef',
+                'name': self._name,
+                'fullname': self._fullname,
+                'arguments': [a.serialize() for a in self.arguments],
+                'type': None if self.type is None else self.type.serialize(),
+                'is_class': self.is_class,
+                # TODO: Various other flags
```
Comment (JukkaL) [inline mypy/nodes.py]: Add other flags, or leave this open? If we leave this open, it might be worth adding an issue for this so we won't forget about it.

File: mypy/nodes.py [hunk 14]
```diff
@@ -270,6 +310,20 @@ def name(self) -> str:
     def accept(self, visitor: NodeVisitor[T]) -> T:
         return visitor.visit_overloaded_func_def(self)
 
+    def serialize(self) -> JsonDict:
+        return {'.class': 'OverloadedFuncDef',
```
Comment (JukkaL) [inline mypy/nodes.py]: I wonder if `is_property` and `_fullname` should be preserved (inherited from `FuncBase`). Also, does `info` get set somewhere?

File: mypy/nodes.py [hunk 15]
```diff
@@ -448,6 +560,37 @@ def fullname(self) -> str:
...
         return visitor.visit_var(self)
 
+    def serialize(self) -> JsonDict:
+        # TODO: Leave default values out?
+        data = {'.class': 'Var',
+                'name': self._name,
+                'fullname': self._fullname,
+                'type': None if self.type is None else self.type.serialize(),
+                'is_self': self.is_self,
+                'is_ready': self.is_ready,  # TODO: is this needed?
```
Comment (JukkaL) [inline mypy/nodes.py]: This is not deserialized. However, for fully type checked files this should always be true.

File: mypy/nodes.py [hunk 16]
```diff
@@ -486,6 +629,30 @@ def accept(self, visitor: NodeVisitor[T]) -> T:
     def is_generic(self) -> bool:
         return self.info.is_generic()
 
+    def serialize(self) -> JsonDict:
+        # Not serialized: defs, base_type_exprs
```
Comment (JukkaL) [inline mypy/nodes.py]: Also, `decorators` is not serialized.

File: mypy/nodes.py [hunk 17]
```diff
@@ -895,12 +1062,30 @@ def __init__(self, name: str) -> None:
         self.name = name
         self.literal_hash = ('Var', name,)
 
-    def type_node(self):
-        return cast(TypeInfo, self.node)
-
     def accept(self, visitor: NodeVisitor[T]) -> T:
         return visitor.visit_name_expr(self)
 
+    def serialize(self) -> JsonDict:
```
Comment (JukkaL) [inline mypy/nodes.py]: When do we need to serialize `NameExpr` objects? I couldn't find any examples where this would be needed, so perhaps it can be removed.

File: mypy/nodes.py [hunk 18]
```diff
@@ -1604,16 +1811,19 @@ def calculate_mro(self) -> None:
 
         Raise MroError if cannot determine mro.
         """
-        self.mro = linearize_hierarchy(self)
+        mro = linearize_hierarchy(self)
+        if mro is None: return  # TODO: Or raise MroError()?
```
Comment (JukkaL) [inline mypy/nodes.py]: Raising `MroError` seems better than just silently failing. We may have to process it somewhere, but if we have a serialized AST, the MRO should always be good.

File: mypy/nodes.py [hunk 19]
```diff
@@ -1649,6 +1859,46 @@ def __str__(self) -> str:
                             ('Names', sorted(self.names.keys()))],
                            'TypeInfo')
 
+    def serialize(self) -> Union[str, JsonDict]:
+        data = {'.class': 'TypeInfo',
+                'fullname': self.fullname(),
+                'alt_fullname': self.alt_fullname,
+                'subtypes': [t.serialize() for t in self.subtypes],
```
Comment (JukkaL) [inline mypy/nodes.py]: We shouldn't serialize `TypeInfo`s for subtypes but include references. However, it seems that this isn't actually used, so we may be able to just leave it out and remove the `subtypes` attribute.

File: mypy/nodes.py [hunk 20]
```diff
@@ -1649,6 +1859,46 @@ def __str__(self) -> str:
...
+                'names': self.names.serialize(self.alt_fullname or self.fullname()),
+                'defn': self.defn.serialize(),
```
Comment (JukkaL) [inline mypy/nodes.py]: Maybe add comment mentioning that this is where all `ClassDef` objects come from so there shouldn't be duplicates.

File: mypy/nodes.py [hunk 21]
```diff
@@ -1714,6 +1967,57 @@ def __str__(self) -> str:
...
+        data = {'.class': 'SymbolTableNode',
+                'kind': node_kinds[self.kind],
+                }  # type: JsonDict
+        if self.tvar_id:
+            data['tvar_id'] = self.tvar_id
+        if not self.module_public:
+            data['module_public'] = False
+        if self.kind == MODULE_REF:
+            if self.node is None:
+                print("*** Missing module cross ref in %s for %s" % (prefix, name))
```
Comment (JukkaL) [inline mypy/nodes.py]: Print statement. Replace with exception?

File: mypy/nodes.py [hunk 22]
```diff
@@ -1714,6 +1967,57 @@ def __str__(self) -> str:
...
+                        fullname = self.node.alt_fullname or self.node.fullname()
+                    else:
+                        fullname = self.node.fullname()
+                    if (fullname is not None and '.' in fullname and
+                            fullname != prefix + '.' + name):
+                        data['cross_ref'] = fullname
+                        return data
+                data['node'] = self.node.serialize()
+            if self.type_override is not None:
+                data['type'] = self.type.serialize()
```
Comment (JukkaL) [inline mypy/nodes.py]: This should be `self.type_override`.

File: mypy/nodes.py [hunk 23]
```diff
@@ -1714,6 +1967,57 @@ def __str__(self) -> str:
...
+        return data
+
+    @classmethod
+    def deserialize(cls, data: JsonDict) -> 'SymbolTableNode':
```
Comment (JukkaL) [inline mypy/nodes.py]: `type_override` is not deserialized.

File: mypy/nodes.py [hunk 24]
```diff
@@ -1714,6 +1967,57 @@ def __str__(self) -> str:
             s += ' : {}'.format(self.type)
         return s
 
+    def serialize(self, prefix: str, name: str) -> JsonDict:
```
Comment (JukkaL) [inline mypy/nodes.py]: Add docstring that describes the arguments.

File: mypy/nodes.py [hunk 25]
```diff
@@ -1714,6 +1967,57 @@ def __str__(self) -> str:
...
+            else:
+                data['cross_ref'] = self.node.fullname()
+        else:
+            if self.node is not None:
+                if prefix is not None:
+                    if isinstance(self.node, TypeInfo):
+                        fullname = self.node.alt_fullname or self.node.fullname()
+                    else:
+                        fullname = self.node.fullname()
+                    if (fullname is not None and '.' in fullname and
```
Comment (JukkaL) [inline mypy/nodes.py]: This is unclear. Add comment.

File: mypy/nodes.py [hunk 26]
```diff
@@ -1732,10 +2036,22 @@ def __str__(self) -> str:
         a[-1] += ')'
         return '\n'.join(a)
 
+    def serialize(self, fullname: str) -> JsonDict:
+        data = {'.class': 'SymbolTable'}  # type: JsonDict
+        for key, value in self.items():
+            if key == '__builtins__':
```
Comment (JukkaL) [inline mypy/nodes.py]: Does this get added somewhere? It would be correct to include `__builtins__`, though probably not very useful. Perhaps add a comment here.

File: mypy/nodes.py [hunk 27]
```diff
@@ -1791,13 +2107,22 @@ class MroError(Exception):
...
     # TODO describe
     if info.mro:
         return info.mro
     bases = info.direct_base_classes()
-    return [info] + merge([linearize_hierarchy(base) for base in bases] +
-                          [bases])
+    lin_bases = []
+    for base in bases:
+        if base is None:
+            print('*** Cannot linearize bases for', info.fullname(), bases)
```
Comment (JukkaL) [inline mypy/nodes.py]: What to do with the print statement?

File: mypy/types.py [hunk 28]
```diff
@@ -197,6 +287,33 @@ def __init__(self, typ: mypy.nodes.TypeInfo, args: List[Type],
...
+    type_ref = None  # type: str
+
+    def serialize(self) -> JsonDict:
+        data = {'.class': 'Instance',
+                }  # type: JsonDict
+        if self.type is not None:
+            data['type_ref'] = self.type.alt_fullname or self.type.fullname()
+        if self.args:
+            data['args'] = [arg.serialize() for arg in self.args]
+        if self.erased:
```
Comment (JukkaL) [inline mypy/types.py]: We don't need to preserve this when serializing, as `erased` only has an effect within expressions, and they aren't serialized. We can default to false always. (Actually, `erased` is not actually used right now, but the attribute can still prove to be useful in the future, so I'd rather preserve it in the full AST.)

File: mypy/types.py [hunk 29]
```diff
@@ -197,6 +287,33 @@ def __init__(self, typ: mypy.nodes.TypeInfo, args: List[Type],
...
+    @classmethod
+    def deserialize(cls, data: JsonDict) -> 'Instance':
+        assert data['.class'] == 'Instance'
+        args = []  # type: List[Type]
+        if 'args' in data:
+            args_list = data['args']
+            assert isinstance(args_list, list)
+            args = [Type.deserialize(arg) for arg in args_list]
+        inst = Instance(None, args, erased=data.get('erased', False))
+        if 'type_ref' in data:
```
Comment (JukkaL) [inline mypy/types.py]: It seems that we always need this. Perhaps require this unconditionally.

File: mypy/types.py [hunk 30]
```diff
@@ -378,6 +531,41 @@ def type_var_ids(self) -> List[int]:
...
+        return {'.class': 'CallableType',
+                'arg_types': [(None if t is None else t.serialize())
+                              for t in self.arg_types],
+                'arg_kinds': self.arg_kinds,
+                'arg_names': self.arg_names,
+                'ret_type': self.ret_type.serialize(),
+                'fallback': self.fallback.serialize(),
+                'name': self.name,
+                # We don't serialize the definition (only used for error messages).
+                'variables': [v.serialize() for v in self.variables],
```
Comment (JukkaL) [inline mypy/types.py]: As a potential future optimization, we could leave out everything related to generic functions for non-generic functions.

File: mypy/types.py [hunk 31]
```diff
@@ -253,9 +395,10 @@ class CallableType(FunctionLike):
     arg_types = None  # type: List[Type]  # Types of function arguments
     arg_kinds = None  # type: List[int]   # mypy.nodes.ARG_ constants
     arg_names = None  # type: List[str]   # None if not a keyword argument
-    min_args = 0                    # Minimum number of arguments
-    is_var_arg = False              # Is it a varargs function?
-    ret_type = None  # type:Type    # Return value type
+    min_args = 0                    # Minimum number of arguments; derived from arg_kinds
+    is_var_arg = False              # Is it a varargs function?  derived from arg_kinds
+    ret_type = None  # type: Type   # Return value type
+    fallback = None  # type: Instance
```
Comment (JukkaL) [inline mypy/types.py]: Repeated from base class `FunctionLike`. I wonder why mypy doesn't complain about this.

**Comment by JukkaL** [comment]
Here's comments for `build.py`, since github doesn't want to show it inline.

* L160 (`# single "main file" Fix this.`): Missing period before "Fix".
* L595 (`get_cache_names`): Add docstring
* L603 (`get_cache_meta`): Add docstirng
* L633 (`# It's a match on (id,...`): Add TODO about build options here. We need to consider `--silent-imports` at least.
* L646 (`write_cache`): Add docstring
* L677 (`write_cache`): Is there a potential race on Windows (`rename` is not necessarily atomic, if I remember correctly)?
* L682 (`write_cache`): Dead code at the end of function. Remove or refactor.
* L859 (`we have a cycle of over 500 modules in the server repo`): The comment won't make sense for anybody outside Dropbox, maybe rephrase such as "Dropbox has a codebase with a cycle of over 500 modules" or "we know of a production codebase with ...".
* L895 (`State`): `# tyoe:` should be `# type:`. Add more comments (to `dep_line_map`, `ancestors`, `caller_state`, `caller_line`).
* L1174 (`load_graph`): Maybe use keyword arguments for `State` to make it more obvious where `None` arguments go. Also elsewhere in this function.
* L1192 (`process_graph`): Add comment that we process from leafs to roots (first imported module, then the module that imports).
* L1197 (`process_graph`): Also explain why we move builtins to be last.
* L1201 (`process_graph`): Around this line, perhaps mention that because of the topographical sort, we don't need to look at dependencies recursively for staleness -- the immediate dependencies are sufficient.
* L1228 (`process_graph`): Maybe mention that *some* environments only have 1-sec time granularity (ext4 on Linux has a finer granularity, for example).
* L1282 (`sorted_components`): The docstring leaves the order of sorting vague. Make it more explicit (from leafs to roots?).
* L1303 (`sorted_components`): Typo in comment: `Exampes:`.
* L1355 (`topsort`): Describe argument and return value in more detail.

**Comment by JukkaL** [comment]
Also L1089 in `build.py`: Print statements. Does the comment mean that a file was edited during a single type checking run?

Some general comments and questions:

* Why are we using BFS instead of DFS? I believe that the choice may change the order in which we process files within cycles. DFS might be closer to what mypy currently does, but not sure.
* It might be worth serializing line number of at least some things such as `Var` nodes, function definitions and classes. These could be useful when generating error messages and shouldn't be expensive.
* Another optimization idea (not sure if I mentioned this before): The entire `Argument` nodes may be redundant and we might not need to serialize them. This should be a pretty nice saving if it works out. Not needed for this PR, though.

One more thing that I'd like to see is at least a few test cases, and an easy way of writing additional test cases. The code is pretty subtle and the serialization code is somewhat fragile. First, you could add a few unit tests for some functions such as toposort in `build.py`. That should be trivial. Second, we could add a new class of type checker tests that use incremental type checking. Here's a sketch of one potential way to do this:

* Create a new test name suffix for incremental type check tests, such as `_Incremental`, and have them live next to type checker tests. Alternatively, create a new kind of test for these. In either approach the test cases should probably live in ` .test` file.
* Some files in a test case can have two versions: original and modified. For modified files, include two copies of the file in a test case description, as we'd type check the files twice. For the second pass copies, we could use a file name such as `foo_pass2.py` or `foo_modified.py` (`foo.py` would be the original version).
* Run type checking twice if we have an incremental test case, and use incremental mode.
  * First type check with empty cache and verify that no errors are reported.
  * Verify that we have cached data for all files.
  * Then replace a subset of files with modified files (`foo_pass2.py` -> `foo.py`). Only do this for files that have second variants.
  * Type check again, using cache, and verify that now we get the expected errors from mypy.
 * Bonus: Verify which files get re-type-checked.
   * Make it possible to declare additional expected type checked files (via dependencies). For example, we can use a comment like `# additional: x.py` in the test program.
    * Record the type checked files that are not using cache data on the second iteration.
    * Verify that we fully type checked exactly the correct set of files (modified files + additional files).
    * Maybe also verify that we didn't generate new cached data for files with errors.

**Comment by JukkaL** [comment]
I'm now finished my first review pass.

**Comment by gvanrossum** [comment]
Awesome! Looks like you caught quite a few things (and found a lot of
prints :-). Hopefully I can add finishing this as one of our OKRs for the
next half-sprint.

**Comment by gvanrossum** [comment]
Hi @JukkaL, I've pushed a new version that incorporates my responses to all your feedback except for the long last one (starting with "Also L1089 in build.py"). I decided not to write individual "done / no, here's why" responses but instead just do everything or explain what's going on in comments.

Next I'm going to work on testing.

**Comment by gvanrossum** [comment]
Jukka, I now have a decent start on the unittest infrastructure.

- There's a new programmatic test module, testgraph.py, that has some tests for topsort() and friends -- it's easy to add more as long as the tests don't require too much setup.

- There's also a new mechanism for data-driven tests of incremental mode. This is implemented in testcheck.py; the new behavior (roughly what you suggested) is triggered either by a test filename containing 'incremental' or by a test name containing 'Incremental'. (We can change the exact conditions easily.) The mechanism runs each such test once with a cold cache expecting no errors, then again with a warn cache after updating files for which a ".next" version is found, expecting the errors from the test's `[out]` section. The code also checks that valid cache files exist exactly for those modules that don't have any errors (after each run).

- I've only written three very basic data-driven tests for now, but it should be easy to add more.

**Comment by gvanrossum** [comment]
So I think I have addressed everything in your review now except

- Serializing line numbers (I haven't found the need so far)
- Your "Bonus" bullet for the test architecture (expecting specific files to be rechecked)
- The answer on BFS vs. DFS -- I added a reminder comment, but I'm not even sure the old code did DFS
- Adding more extensive tests (but you didn't really ask for that -- you said "a few" and I've got that :-)

I think I'd like to leave all those unaddressed.

File: mypy/nodes.py [hunk 32]
```diff
@@ -630,7 +653,7 @@ def is_generic(self) -> bool:
         return self.info.is_generic()
 
     def serialize(self) -> JsonDict:
-        # Not serialized: defs, base_type_exprs
+        # Not serialized: defs, base_type_exprs,d ecorators
```
Comment (JukkaL) [inline mypy/nodes.py]: typo: `d ecorators`

File: mypy/build.py [hunk 33]
```diff
@@ -1353,6 +1367,18 @@ def strongly_connected_components_path(vertices: Set[str],
                                        edges: Dict[str, List[str]]) -> Iterator[Set[str]]:
     """Compute Strongly Connected Components of a graph.
 
+    The graph is a DAG.
```
Comment (JukkaL) [inline mypy/build.py]: What does this mean? The resulting graph where nodes are strongly connected components is a graph?

Comment (gvanrossum) [inline mypy/build.py]: You're reviewing an out-of-date version of this file. When updating the docstrings I realized that this is nonsense and corrected it.

File: mypy/test/testcheck.py [hunk 34]
```diff
@@ -80,17 +114,66 @@ def run_test(self, testcase):
...
-            testcase.output, a,
+            output, a,
             'Invalid type checker output ({}, line {})'.format(
                 testcase.file, testcase.line))
 
+        if incremental and res:
+            self.verify_cache(module_name, program_name, a, res.manager)
+
+    def verify_cache(self, module_name: str, program_name: str, a: List[str],
+                     manager: build.BuildManager):
```
Comment (JukkaL) [inline mypy/test/testcheck.py]: No return type annotation

Comment (gvanrossum) [inline mypy/test/testcheck.py]: Already fixed in a newer version.

**Comment by JukkaL** [comment]
OK, looks good now! Feel free to merge whenever you like. After merging I suggest filing some new issues (for 0.3.2, I assume):

* Add extensive tests for incremental type checking. I'd like more tests before turning incremental mode on by default (+ more manual testing).
* Check which files were rechecked in incremental mode test cases (and which weren't).

I'm not sure either sure whether the old code used DFS, but I have a vague feeling that it might.

**Comment by gvanrossum** [comment]
Filed: #1349 (add tests), #1350 (check which files we rechecked).
